shark_turbine/kernel/wave/decompose_reduce_ops.py

```diff
@@ -0,0 +1,114 @@
+    binary_fn: Callable, src: fx.Node, graph: fx.Graph, subgroup_size: int
+) -> fx.Node:
+    init = src
+    num_steps = int(math.log2(float(subgroup_size)))
+    for i in range(num_steps):
+        shuffle_offset = 2**i
+        shuffle_val = ShuffleOp(init, shuffle_offset, subgroup_size)
+        shuffle_node = get_graph_node(shuffle_val, graph)
+        init = get_graph_node(binary_fn(init, shuffle_node), graph)
+    return init
+
+
+def decompose_reduce_ops(
+    trace: CapturedTrace,
+    constraints: list[Constraint],
+    index_map: dict[IndexSymbol, int],
+):
+    """
+    The lowering for multi_reduction is done in two steps:
+      1. Local Reduce: Each thread reduces all elements carried by it along
+         the reduction dimensions.
+      2. Thread Reduce: Each thread reduces result of step 1 across threads
+         by doing a butterfly shuffle.
+      3. Accumulator Reduce: Each thread reduces it's intermediate reduced
+         results with the accumulator it holds.
+    """
+    # Get reducte nodes.
+    reduce_nodes = trace.walk(lambda node: isinstance(get_custom(node), ReduceOp))
+    if not reduce_nodes:
+        return
+
+    # Setup constraints
+    hardware_constraint = next(
+        c for c in constraints if isinstance(c, HardwareConstraint)
+    )
+    constraint_tile_size = {
+        c.dim: c.tile_size
+        for c in constraints
+        if isinstance(c, TilingConstraint) or isinstance(c, WorkgroupConstraint)
+    }
+    subgroup_size = hardware_constraint.threads_per_wave
+    for node in reduce_nodes:
+        custom = get_custom(node)
+        with custom.graph.inserting_before(custom.fx_node):
+            reduction_src, reduction_acc, reduction_dim = node.args
+            binary_fn = TKW_COMBINER[custom.tkw_op_name]
+            if reduction_dim is None:
+                raise ValueError(
+                    "No reduction dim specified, please specify a reduction dim."
+                )
+
+            # Local Reduce
+            if reduction_dim is not custom.type.symbolic_shape[-1]:
+                raise NotImplementedError(
+                    "Only implemented reduction on fastest dimension."
+                )
+            reduction_block_size = constraint_tile_size[reduction_dim]
+            reduction_size = reduction_block_size.subs(index_map)
+            local_reduction_size = reduction_size / subgroup_size
+            local_reduction = emit_local_reduction(
+                binary_fn, reduction_src, custom.graph, local_reduction_size
+            )
+
+            # Global Reduce
+            global_reduction = emit_global_reduction(
+                binary_fn, local_reduction, custom.graph, subgroup_size
+            )
+
+            # Local Accumulator Reduce
+            final_reduction = global_reduction
+            if reduction_acc is not None:
+                final_reduction = get_graph_node(
+                    binary_fn(reduction_acc, global_reduction), graph
```

Comment:
graph here is not defined in this function, should it be custom.graph?

Reply:
Good catch, I'll push up a fix for this :)